# Notebook for How to Win a Data Science Competition: Learn from Top Kagglers Final Project

### Import the data

In [1]:
# The code was removed by Watson Studio for sharing.

# Exploratory Data Analysis

In [2]:
print(item_categories.shape)
item_categories.head()

(84, 2)


item_category_name  item_category_id
0  PC - Гарнитуры/Наушники                 0
1         Аксессуары - PS2                 1
2         Аксессуары - PS3                 2
3         Аксессуары - PS4                 3
4         Аксессуары - PSP                 4

In [3]:
print(items.shape)
items.head()

(22170, 3)


item_name  item_id  \
0          ! ВО ВЛАСТИ НАВАЖДЕНИЯ (ПЛАСТ.)         D        0   
1  !ABBYY FineReader 12 Professional Edition Full...        1   
2      ***В ЛУЧАХ СЛАВЫ   (UNV)                    D        2   
3    ***ГОЛУБАЯ ВОЛНА  (Univ)                      D        3   
4        ***КОРОБКА (СТЕКЛО)                       D        4   

   item_category_id  
0                40  
1                76  
2                40  
3                40  
4                40

In [4]:
print(sales_train.shape)
sales_train.head()

(2935849, 6)


date  date_block_num  shop_id  item_id  item_price  item_cnt_day
0  02.01.2013               0       59    22154      999.00           1.0
1  03.01.2013               0       25     2552      899.00           1.0
2  05.01.2013               0       25     2552      899.00          -1.0
3  06.01.2013               0       25     2554     1709.05           1.0
4  15.01.2013               0       25     2555     1099.00           1.0

In [5]:
print(shops.shape)
shops.head()

(60, 2)


shop_name  shop_id
0   !Якутск Орджоникидзе, 56 фран        0
1   !Якутск ТЦ "Центральный" фран        1
2                Адыгея ТЦ "Мега"        2
3  Балашиха ТРК "Октябрь-Киномир"        3
4        Волжский ТЦ "Волга Молл"        4

In [6]:
print(test.shape)
test.head()

(214200, 3)


ID  shop_id  item_id
0   0        5     5037
1   1        5     5320
2   2        5     5233
3   3        5     5232
4   4        5     5268

In [7]:
shops["shop_id"].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59])

### There are 60 unique shop ids.

In [8]:
shops.describe()

shop_id
count  60.000000
mean   29.500000
std    17.464249
min     0.000000
25%    14.750000
50%    29.500000
75%    44.250000
max    59.000000

In [9]:
items.head()

item_name  item_id  \
0          ! ВО ВЛАСТИ НАВАЖДЕНИЯ (ПЛАСТ.)         D        0   
1  !ABBYY FineReader 12 Professional Edition Full...        1   
2      ***В ЛУЧАХ СЛАВЫ   (UNV)                    D        2   
3    ***ГОЛУБАЯ ВОЛНА  (Univ)                      D        3   
4        ***КОРОБКА (СТЕКЛО)                       D        4   

   item_category_id  
0                40  
1                76  
2                40  
3                40  
4                40

In [10]:
items.describe()

item_id  item_category_id
count  22170.00000      22170.000000
mean   11084.50000         46.290753
std     6400.07207         15.941486
min        0.00000          0.000000
25%     5542.25000         37.000000
50%    11084.50000         40.000000
75%    16626.75000         58.000000
max    22169.00000         83.000000

In [11]:
items["item_id"].unique().shape[0]

22170

### There are 22,170 unique item category ids. Not possible to list!

In [12]:
sales_train.head()

date  date_block_num  shop_id  item_id  item_price  item_cnt_day
0  02.01.2013               0       59    22154      999.00           1.0
1  03.01.2013               0       25     2552      899.00           1.0
2  05.01.2013               0       25     2552      899.00          -1.0
3  06.01.2013               0       25     2554     1709.05           1.0
4  15.01.2013               0       25     2555     1099.00           1.0

In [13]:
sales_train.describe()

date_block_num       shop_id       item_id    item_price  item_cnt_day
count    2.935849e+06  2.935849e+06  2.935849e+06  2.935849e+06  2.935849e+06
mean     1.456991e+01  3.300173e+01  1.019723e+04  8.908532e+02  1.242641e+00
std      9.422988e+00  1.622697e+01  6.324297e+03  1.729800e+03  2.618834e+00
min      0.000000e+00  0.000000e+00  0.000000e+00 -1.000000e+00 -2.200000e+01
25%      7.000000e+00  2.200000e+01  4.476000e+03  2.490000e+02  1.000000e+00
50%      1.400000e+01  3.100000e+01  9.343000e+03  3.990000e+02  1.000000e+00
75%      2.300000e+01  4.700000e+01  1.568400e+04  9.990000e+02  1.000000e+00
max      3.300000e+01  5.900000e+01  2.216900e+04  3.079800e+05  2.169000e+03

In [14]:
submission.head()

ID  item_cnt_month
0   0             0.5
1   1             0.5
2   2             0.5
3   3             0.5
4   4             0.5

In [15]:
test.head()

ID  shop_id  item_id
0   0        5     5037
1   1        5     5320
2   2        5     5233
3   3        5     5232
4   4        5     5268

## Restructure train data to match test data

In [16]:
import numpy as np
from itertools import product
index_cols = ['shop_id', 'item_id', 'date_block_num']

# Iterate through all shops/items combinations
restructured_data = []
for block_num in sales_train['date_block_num'].unique():
    restructured_shops = sales_train.loc[sales_train['date_block_num'] == block_num, 'shop_id'].unique()
    restructured_items = sales_train.loc[sales_train['date_block_num'] == block_num, 'item_id'].unique()
    restructured_data.append(np.array(list(product(*[restructured_shops, restructured_items, [block_num]])),dtype='int32'))
restructured_data = pd.DataFrame(np.vstack(restructured_data), columns = index_cols,dtype=np.int32)

In [17]:
restructured_data.head()

shop_id  item_id  date_block_num
0       59    22154               0
1       59     2552               0
2       59     2554               0
3       59     2555               0
4       59     2564               0

### There are 50% missing data

In [18]:
test.merge(sales_train, how='left', on=['shop_id', 'item_id']).isnull().sum()


ID                     0
shop_id                0
item_id                0
date              102796
date_block_num    102796
item_price        102796
item_cnt_day      102796
dtype: int64

### we drop the outliers

In [19]:
sales_train = sales_train[sales_train.item_price<100000]
sales_train = sales_train[sales_train.item_cnt_day<=900]

## We have to make predictions by months

In [20]:
sales_train_combined = sales_train.groupby(['date_block_num','shop_id','item_id']).agg({'item_cnt_day': 'sum','item_price': np.mean}).reset_index()

## Fill NA entries with 0

In [21]:
sales_train_combined = pd.merge(restructured_data,sales_train_combined,on=['date_block_num','shop_id','item_id'],how='left').fillna(0)

## Merge remaining columns item_id and item_category_id

In [22]:
sales_train_combined = pd.merge(sales_train_combined,items,on=['item_id'],how='left')
sales_train_combined = pd.merge(sales_train_combined,item_categories, on=['item_category_id'], how='left')

In [23]:
sales_train_combined.head()

shop_id  item_id  date_block_num  item_cnt_day  item_price  \
0       59    22154               0           1.0       999.0   
1       59     2552               0           0.0         0.0   
2       59     2554               0           0.0         0.0   
3       59     2555               0           0.0         0.0   
4       59     2564               0           0.0         0.0   

                                           item_name  item_category_id  \
0                                  ЯВЛЕНИЕ 2012 (BD)                37   
1           DEEP PURPLE  The House Of Blue Light  LP                58   
2           DEEP PURPLE  Who Do You Think We Are  LP                58   
3            DEEP PURPLE 30 Very Best Of 2CD (Фирм.)                56   
4  DEEP PURPLE Perihelion: Live In Concert DVD (К...                59   

                    item_category_name  
0                       Кино - Blu-Ray  
1                       Музыка - Винил  
2                       Музыка - Винил  
3  Музыка - CD фирменного производства  
4           Музыка - Музыкальное видео

# Mean encoding

In [24]:
for type_id in ['item_id','shop_id','item_category_id']:
    for column_id, agg_func, func_name in [('item_price',np.mean,'avg'),('item_cnt_day',np.sum,'sum'),('item_cnt_day',np.mean,'avg')]:

        mean_df = sales_train_combined.groupby([type_id,'date_block_num']).agg(agg_func).reset_index()[[column_id,type_id,'date_block_num']]
        mean_df.columns = [type_id+'_'+func_name+'_'+column_id, type_id,'date_block_num']
        
        sales_train_combined = pd.merge(sales_train_combined,mean_df,on=['date_block_num', type_id],how='left')

### Attempt to add additional time delayed feature, for this month we are training our data we have previous months data to aid in training

In [25]:
temp_test = test.copy()
temp_test['date_block_num'] = 34
temp_test.drop('ID', axis=1, inplace=True)

temp_test = temp_test.merge(items, how='left', on='item_id')
temp_test = temp_test.merge(item_categories, how='left', on='item_category_id')
temp_test.drop('item_name', axis=1, inplace=True)

sales_train_combined.drop('item_name', axis=1, inplace=True)
sales_train_combined = pd.concat([sales_train_combined,temp_test], axis=0, ignore_index=True)

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:10: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.



In [26]:
additional_features = ['item_id_avg_item_price',
'item_id_sum_item_cnt_day',
'item_id_avg_item_cnt_day',
'shop_id_avg_item_price',
'shop_id_sum_item_cnt_day',
'shop_id_avg_item_cnt_day',
'item_category_id_avg_item_price',
'item_category_id_sum_item_cnt_day',
'item_category_id_avg_item_cnt_day',
'item_cnt_day']

#features = [1]
features = [1, 2, 3, 4, 5, 6, 9, 12]
for feature in features:
    sales_new_df = sales_train_combined.copy()
    sales_new_df.date_block_num+=feature
    sales_new_df = sales_new_df[['date_block_num','shop_id','item_id']+additional_features]
    sales_new_df.columns = ['date_block_num','shop_id','item_id']+ [delayed_feat+'_delay_'+str(feature) for delayed_feat in additional_features]
    sales_train_combined = sales_train_combined.merge(sales_new_df,on=['date_block_num','shop_id','item_id'] ,how='left')

### Fill missing values with zeros

In [27]:
for feat in sales_train_combined.columns:
    if 'item_cnt' in feat:
        sales_train_combined[feat]=sales_train_combined[feat].fillna(0)
    elif 'item_price' in feat:
        sales_train_combined[feat]=sales_train_combined[feat].fillna(sales_train_combined[feat].median())

In [28]:
cols_to_drop = additional_features[:-1] + ['item_price'] + ['item_category_name']
sales_train_combined = sales_train_combined[sales_train_combined['date_block_num']>12]

# Validation

In [29]:
X_train = sales_train_combined[sales_train_combined['date_block_num']<33].drop(cols_to_drop, axis=1)
X_cv =  sales_train_combined[sales_train_combined['date_block_num']==33].drop(cols_to_drop, axis=1)
X_test = sales_train_combined[sales_train_combined['date_block_num']==34].drop(cols_to_drop, axis=1)

# Model Training

In [30]:
params = {
        'eta': 0.08, #best 0.08
        'max_depth': 7,
        'objective': 'reg:linear',
        'eval_metric': 'rmse',
        'seed': 3,
        'gamma':1,
        'silent': True
    }

cols = [c for c in X_train.columns if c not in ['date_block_num', 'item_cnt_day',]]

### We make use of XGBoost from scikit learn

In [32]:
import numpy as np
import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 100)

from itertools import product
from sklearn.preprocessing import LabelEncoder

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

from xgboost import XGBRegressor
from xgboost import plot_importance

def plot_features(booster, figsize):    
    fig, ax = plt.subplots(1,1,figsize=figsize)
    return plot_importance(booster=booster, ax=ax)

import time
import sys
import gc
import pickle

### Separate the features and predictions

In [33]:
X_training = X_train[cols]
Y_training = X_train['item_cnt_day']
X_valid = X_cv[cols]
Y_valid = X_cv['item_cnt_day']

### Start training the model

In [34]:
ts = time.time()

model = XGBRegressor(
    max_depth=8,
    n_estimators=1000,
    min_child_weight=300, 
    colsample_bytree=0.8, 
    subsample=0.8, 
    eta=0.3,    
    seed=42)

model.fit(
    X_training, 
    Y_training, 
    eval_metric="rmse", 
    eval_set=[(X_training, Y_training), (X_valid, Y_valid)], 
    verbose=True, 
    early_stopping_rounds = 10)

time.time() - ts

[04:46:25] Tree method is automatically selected to be 'approx' for faster speed. to use old behavior(exact greedy algorithm on single machine), set tree_method to 'exact'
[0]	validation_0-rmse:3.41929	validation_1-rmse:2.59817
Multiple eval metrics have been passed: 'validation_1-rmse' will be used for early stopping.

Will train until validation_1-rmse hasn't improved in 10 rounds.
[1]	validation_0-rmse:3.31998	validation_1-rmse:2.51077
[2]	validation_0-rmse:3.24611	validation_1-rmse:2.4445
[3]	validation_0-rmse:3.17077	validation_1-rmse:2.38148
[4]	validation_0-rmse:3.11136	validation_1-rmse:2.33204
[5]	validation_0-rmse:3.05498	validation_1-rmse:2.28333
[6]	validation_0-rmse:3.00569	validation_1-rmse:2.24458
[7]	validation_0-rmse:2.9651	validation_1-rmse:2.20916
[8]	validation_0-rmse:2.92949	validation_1-rmse:2.1784
[9]	validation_0-rmse:2.8976	validation_1-rmse:2.15185
[10]	validation_0-rmse:2.868	validation_1-rmse:2.13062
[11]	validation_0-rmse:2.84404	validation_1-rmse:2.11349
[

3323.371490240097

In [35]:
Y_pred = model.predict(X_valid).clip(0, 20)
Y_test = model.predict(X_test[cols]).clip(0, 20)

submission = pd.DataFrame({
    "ID": test.index, 
    "item_cnt_month": Y_test
})
submission.to_csv('submission.csv', index=False)

# save predictions for an ensemble
pickle.dump(Y_pred, open('xgb_train.pickle', 'wb'))
pickle.dump(Y_test, open('xgb_test.pickle', 'wb'))

# Saving the Data !

In [37]:
# this portion only applies to Watson Studio
from project_lib import Project
project = Project(None, '7dcfcd13-e22e-4d47-b4f4-32bb024bc8e3', 'p-36f61474aa06cf9db5dbffd3a4ea2877ce6ef560')
pc = project.project_context

file_data=open('submission.csv', 'rb')
project.save_data(data=file_data.read(),file_name='submission_full.csv',overwrite=True)
file_data.close()    

In [41]:
# this portion only applies to Watson Studio
# save pickle
project.save_data(data=pickle.dumps('xgb_train.pickle'),file_name='RF_train.pk',overwrite=True)
project.save_data(data=pickle.dumps('xgb_test.pickle'),file_name='RF_test.pk',overwrite=True)

{'file_name': 'RF_test.pk',
 'message': 'File saved to project storage.',
 'bucket_name': 'competitivedatascience-donotdelete-pr-4q7ovrnazzvi09',
 'asset_id': 'cb18a78c-40b3-42d6-97fa-e6ba82ec1c8a'}

In [ ]:
# this portion only applies to Watson Studio
# Download and load model
#pickle.load(project.get_file('RF.pkl'))